chore(docs) document Kafka ILP connector (#1165)
questdb/questdb.io: docs/third-party-tools/kafka/ilp.md

QuestDB table, `example_table`.

Create a file `config/questdb-connector.properties` with the following contents:

```shell title="Create a configuration file"
name=questdb-sink
connector.class=io.questdb.kafka.QuestDBSinkConnector
host=localhost:9009
topics=example-topic
table=example_table
include.key=false
value.converter=org.apache.kafka.connect.json.JsonConverter
value.converter.schemas.enable=false
key.converter=org.apache.kafka.connect.storage.StringConverter
```

### Start Kafka

The commands listed in this section must be run from the Kafka home directory
and in the order shown below.

1. Start the Kafka Zookeeper used to coordinate the server:

```shell
bin/zookeeper-server-start.sh  config/zookeeper.properties
```

2. Start an Apache Kafka server:

```shell
bin/kafka-server-start.sh  config/server.properties
```

3. Start the Kafka Connect connector:

```shell
bin/connect-standalone.sh config/connect-standalone.properties config/questdb-connector.properties
```

### Publish messages

Messages can be published via the console producer script:

```shell
bin/kafka-console-producer.sh --topic example-topic --bootstrap-server localhost:9092
```

> A `>` greater-than symbol indicates that a message can be published to the

Reviewer: ```suggestion
A greater-than symbol, `>`, indicates that a message can be published to the
```

QuestDB pipeline! 🎉

### Additional sample projects

You can find additional sample projects on
the [QuestDB Kafka connector](https://github.com/questdb/kafka-questdb-connector/tree/main/kafka-questdb-connector-samples)
Github project page. It includes
a [sample integration](https://github.com/questdb/kafka-questdb-connector/tree/main/kafka-questdb-connector-samples/stocks)
with [Debezium](https://debezium.io/) for [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture) from
Postgres.

## Configuration manual
This section lists configuration options as well as further information about the Kafka Connect connector.
### Configuration Optons

The connector supports following configuration options:

| Name                              | Type    | Example                                                     | Default            | Meaning                                                    |
|-----------------------------------|---------|-------------------------------------------------------------|--------------------|------------------------------------------------------------|

> | topics                            | STRING  | orders                                                      | N/A                | Topics to read from                                        |

Reviewer: I don't think we use capital letters for data type elsewhere. I think `string` would do?

---
title: ILP connector
description: QuestDB ships a Kafka Connect connector for ingesting messages from a Kafka via the ILP protocol.
---

QuestDB ships [Kafka Connect connector](https://github.com/questdb/kafka-questdb-connector) for fast ingestion from
Kafka into QuestDB.

This page has two parts. The first part is a practical guide showing how to integrate the QuestDB Kafka Connector
with Apache Connect. The second part is more a reference for the connector, it lists all configuration
options and discusses common pitfalls. If you are not interested in the guide then you can jump right into
the [reference part.](#reference-documentation)

## Quick Guide
We start with a practical guide showing how to use the ILP connector to read JSON data from Kafka topic and write them
as rows into a QuestDB table.
### Prerequisites

You will need the following:

> - Kafka

Author: Apache Kafka is the full name and trademark. Surely the Apache Foundation (the trademark owner) would prefer to use the full name exclusively. I feel it's too repetitive and the project is colloquially called just "Kafka". I'm good with either option, up to you, dear editor :)

> - Kafka

Reviewer: Agree we can drop Apache. Reading through this again I wonder if   we should be consistent in using Kafka or Kafka Connect? (Cf L10)

> - Kafka

Author: it should be just Kafka. 
Kafka Connect is part of (Apache) Kafka distribution.

| firstname <sub>string</sub> | lastname <sub>string</sub> | age <sub>long</sub> | address_street <sub>string</sub> | address_city <sub>string</sub> |
|-----------------------------|----------------------------|---------------------|----------------------------------|--------------------------------|
| John                        | Doe                        | 30                  | Main Street                      | New York                       |

### Supported serialization formats

The connector does not do data deserialization on its own. It relies on Kafka Connect converters to deserialize data.
It's been tested predominantly with JSON, but it should work with any converter, including Avro. Converters can be
configured using `key.converter` and `value.converter` options, see the table above.

### Designated Timestamps

The connector supports [designated timestamps.](https://questdb.io/docs/concept/designated-timestamp/) If the message
contains a field with a timestamp, the connector can use it as a timestamp for the row. The field name must be
configured using `timestamp.field.name` option. The field must either a simple number or a timestamp. When it's a simple

> number, the connector will interpret it as a Unix timestamp in milliseconds.

Author: this now says:
> When the field is set to an integer the connector will interpret it as a Unix timestamp in milliseconds.

### QuestDB Symbol Type

QuestDB supports a special type called [Symbol](https://questdb.io/docs/concept/symbol/). Use the `symbols`
configuration option to specify which columns should be created as the `symbol` type

### Target Table Considerations

> When a target table does not exist in QuestDB then it will be automatically created when a first row arrives. This is
> recommended approach for development and testing.
> 

Reviewer: Should "arrives" be changed to "is committed" to be more accurate? Is this correct at all?

```suggestion
When a target table does not exist in QuestDB, it will be automatically created when the first row arrives. This approach is only recommended for development and testing.
```

Author: I will double-check, but I believe the target table is created even before ILP data are physically committed.

Reviewer: TBD as "arrives" is vague - as a user, I don't know if it means the row will be visible to me

> In production, it's recommended to [create tables manually via SQL](https://questdb.io/docs/reference/sql/create-table/)
> . This gives you more control over the table schema, allow per-table partitioning, creating indexes, etc.

Reviewer: ```suggestion
In production, it's recommended to use the SQL [CREATE TABLE](https://questdb.io/docs/reference/sql/create-table/) keyword, because it gives you more control over the table schema, allowing per-table partitioning, creating indexes, etc.
```

> ### FAQ

Reviewer: ```suggestion
## FAQ
```

**Q**: Does this connector work with Schema Registry?  

> **A**: The Connector does not care about serialization strategy used. It relies on Kafka Connect converters to

Reviewer: I would avoid saying 'does not care about' but not sure if this is correct:

```suggestion
**A**: The Connector works independently of the serialization strategy used. It relies on Kafka Connect converters to
```

Author: you suggestion is correct, thanks!

Reviewer: TBD this has not been incorporated.

deserialize data. Converters can be configured using `key.converter` and `value.converter` options, see the
configuration section above.

**Q**: I'm getting this
error: `org.apache.kafka.connect.errors.DataException: JsonConverter with schemas.enable requires "schema" and "payload" fields and may not contain additional fields. If you are trying to deserialize plain JSON data, set schemas.enable=false in your converter configuration.`    
**A**: This error means that the connector is trying to deserialize data using a converter that expects a schema. The

> connector does not require schemas, so you need to configure the converter to not expect a schema. For example, if you
> are using JSON converter, you need to set `value.converter.schemas.enable=false` or `key.converter.schemas.enable=false`

Reviewer: ```suggestion
connector does not require schemas, so you need to configure the converter to not expect a schema. For example, if you
are using a JSON converter, you need to set `value.converter.schemas.enable=false` or `key.converter.schemas.enable=false`
```

in the connector configuration.

**Q**: Does this connector work with Debezium?  
**A**: Yes, it's been tested with Debezium as a source and
a [sample project](https://github.com/questdb/kafka-questdb-connector/tree/main/kafka-questdb-connector-samples/stocks)

> is available. Bear in mind that QuestDB is meant to be used as append-only database hence updates should be translated

Reviewer: ```suggestion
is available. Bear in mind that QuestDB is meant to be used as an append-only database;  hence, updates should be translated
```

as new inserts. The connector supports Debezium's `ExtractNewRecordState` transformation to extract the new state of the

> record. The transform by default drops DELETE events so no need to handle it explicitly.

Reviewer: ```suggestion
record. The transformation by default drops DELETE events, so there is no need to handle them explicitly.
```

> **Q**: QuestDB is a time-series database, how does it fit into Change Data Capture via Debezium?  

Reviewer: ```suggestion
**Q**: QuestDB is a time series database, how does it fit into Change Data Capture via Debezium?  
```

**A**: QuestDB works with Debezium just great! This is the recommended pattern: Transactional applications use a
relational database to store the current state of the data. QuestDB is used to store the history of changes. Example:

> Imagine you have a Postgres table with the most recent stock prices. Whenever a stock price changes an application

Reviewer: ```suggestion
Imagine you have a PostgreSQL table with the most recent stock prices. Whenever a stock price changes an application
```

> updates the Postgres table. Debezium capture each UPDATE/INSERT and pushes it as an event to Kafka. Kafka Connect

Reviewer: ```suggestion
updates the PostgreSQL table. Debezium captures each UPDATE/INSERT and pushes it as an event to Kafka. Kafka Connect
```

> QuestDB connector reads the events and inserts them into QuestDB. This way Postgres will have the most recent stock
> prices and QuestDB will have the history of changes. You can use QuestDB to build a dashboard with the most recent stock

Reviewer: Question:
If the most recent prices are in PostgreSQL, then how is it possible to build a price dashboard using QuestDB? 

```suggestion
QuestDB connector reads the events and inserts them into QuestDB. In this way, PostgreSQL will have the most recent stock
prices and QuestDB will have the history of changes. You can use QuestDB to build a dashboard with the most recent stock
```

**A**: Use the ReplaceField transformation to remove unwanted fields. For example, if you want to remove
the `address` field, you can use the following configuration:

```json
{
  "name": "questdb-sink",
  "config": {
    "connector.class": "io.questdb.kafka.QuestDBSinkConnector",
    "host": "localhost:9009",
    "topics": "Orders",
    "table": "orders_table",
    "key.converter": "org.apache.kafka.connect.storage.StringConverter",
    "value.converter": "org.apache.kafka.connect.json.JsonConverter",
    "transforms": "removeAddress",
    "transforms.removeAddress.type": "org.apache.kafka.connect.transforms.ReplaceField$Value",
    "transforms.removeAddress.blacklist": "address"
  }
}
```

See [ReplaceField documentation](https://docs.confluent.io/platform/current/connect/transforms/replacefield.html#replacefield)
for more details.

**Q**: I need to run Kafka Connect on Java 8, but the connector says it requires Java 11. What should I do?  
**A**: The Kafka Connect-specific part of the connectors works with Java 8. The requirement for Java 11 is coming
from QuestDB client itself. The zip archive contains 2 JARs: `questdb-kafka-connector-VERSION.jar`

> and `questdb-VERSION.jar`. You can use replace the latter with `questdb-VERSION-jdk8.jar` from

Reviewer: ```suggestion
and `questdb-VERSION.jar`. You can replace the latter with `questdb-VERSION-jdk8.jar` from
```

> the [Maven central](https://mvnrepository.com/artifact/org.questdb/questdb/6.5.4-jdk8). Bear in mind this setup is not

Reviewer: ```suggestion
the [Maven central](https://mvnrepository.com/artifact/org.questdb/questdb/6.5.4-jdk8). Please note that this setup is not
```

> description: QuestDB ships a Kafka Connect connector for ingesting messages from a Kafka via the ILP protocol.

Reviewer: A Kafka - a Kafka Connect instance?

Reviewer: - "Kafka connector" using Kafka Connect
- Apache to be mentioned in the first instance and then just Kafka
- ILP connector using Kafka Connect --> QuestDB Kafka connector (Kafka Connect connector)
